integrators/explicit_embedded_runge_kutta_nyström_integrator_body.hpp

```diff
@@ -0,0 +1,204 @@
+  DoublePrecision<Time> t = initial_value.time;
+
+  // Position increment (high-order).
+  std::vector<Displacement> ∆q_hat(dimension);
+  // Velocity increment (high-order).
+  std::vector<Velocity> ∆v_hat(dimension);
+  // Current position.
+  std::vector<DoublePrecision<Position>> q_hat = initial_value.positions;
+  // Current velocity.
+  std::vector<DoublePrecision<Velocity>> v_hat = initial_value.momenta;
+
+  // Difference between the low- and high-order approximations of the position.
+  std::vector<Displacement> q_error_estimate(dimension);
+  // Difference between the low- and high-order approximations of the velocity.
+  std::vector<Velocity> v_error_estimate(dimension);
+
+  // Current Runge-Kutta-Nyström stage.
+  std::vector<Position> q_stage(dimension);
+  // Accelerations at each stage.
+  // TODO(egg): this is a rectangular container, use something more appropriate.
+  std::vector<std::vector<Acceleration>> g(stages_);
+  for (auto& g_stage : g) {
+    g_stage.resize(dimension);
+  }
+
+  bool at_end = false;
+  double tolerance_to_error_ratio;
+
+  // No step size control on the first step.
+  goto runge_kutta_nyström_step;
+
+  while (!at_end) {
+    // Compute the next step with decreasing step sizes until the error is
+    // tolerable.
+    do {
+      // Adapt step size.
+      h *= safety_factor * std::pow(tolerance_to_error_ratio,
+                                    1.0 / (lower_order_ + 1));
+
+    runge_kutta_nyström_step:
+      // Termination condition.
+      Time const time_to_end = t_final - t.value - t.error;
+      at_end = forward && h >= time_to_end || !forward && h <= time_to_end;
+      if (at_end) {
+        // The chosen step size will overshoot.  Clip it to just reach the end,
+        // and terminate if the step is accepted.
+        h = time_to_end;
+      }
+
+      // Runge-Kutta-Nyström iteration; fills |g|.
+      for (int i = 0; i < stages_; ++i) {
+        Time const t_stage = t.value + c_[i] * h;
+        for (int k = 0; k < dimension; ++k) {
+          Acceleration ∑_a_ij_g_j;
+          for (int j = 0; j < i; ++j) {
+            ∑_a_ij_g_j += a_[i][j] * g[j][k];
+          }
+          q_stage[k] = q_hat[k].value +
+                           h * (c_[i] * v_hat[k].value + h * ∑_a_ij_g_j);
+        }
+        compute_acceleration(t_stage, q_stage, &g[i]);
+      }
+      // TODO(egg): handle the FSAL case.
+
+      // Increment computation and step size control.
+      for (int k = 0; k < dimension; ++k) {
+        Acceleration ∑_b_hat_i_g_i;
+        Acceleration ∑_b_i_g_i;
+        Acceleration ∑_b_prime_hat_i_g_i;
+        Acceleration ∑_b_prime_i_g_i;
+        // Low-order position increment.
+        Displacement ∆q_k;
+        // Low-order velocity increment.
+        Velocity ∆v_k;
+        // Please keep eight assigments below aligned, they become illegible
+        // otherwise.
+        for (int i = 0; i < stages_; ++i) {
+          ∑_b_hat_i_g_i       += b_hat_[i] * g[i][k];
```

Comment:
Well, it's the *k*th entry of ∑ *b̂*<sub>*i*</sub> ***g***<sub>*i*</sub>.
Like `∑_a_ij_g_j` above, the whole thing is in a loop over `k`, but I'm not sure adding the k makes it clearer (in particular when comparing with textbooks/papers, where ∑ *b̂*<sub>*i*</sub> ***g***<sub>*i*</sub> appears in that form rather than entrywise).